[FIX] Adapt tutorial to new bitvector class. (#10)
seqan/learning-resources: doc/nanocourse/bitvectors.md

As noted above we need to represent bits using larger entities available in the C++ language.
It usually pays off if you choose an integer with the size of a machine word `w`, which is 64 on modern architectures.

> Therefore we will use a the C++ type `uint64_t` for our representation.

Reviewer: ```suggestion
Therefor, we will use the C++ type `uint64_t` for our representation.
```

\assignment{A compact bitvector}

Implement a new data structure called `Bitvector` (a `struct`) that has a member variable called `data` of type
`std::vector` over `uint64_t`. The constructor should have one parameter that is the number of bits you would

> like to store in you bitvector.

Reviewer: ```suggestion
like to store in your bitvector.
```

> If you arn't an experienced C++ programmer, take a look at this backbone and fill in the *TODOs*:

Reviewer: ```suggestion
If you aren't an experienced C++ programmer, take a look at this backbone and fill in the *TODOs*:
```

The block and superblock values obviously need to be stored using an arithmetic data type
(e.g. `uint8_t`, `uint16_t`, `uint32_t`, `uint64_t`).
Given an arbitrary bitvector of length n, word length `w = 64 bits`
and a superblock length of `s = 64 * 25 = 1600 bits`,

> Which type would you choose for superblock entries, and which type for block entries and why?

Reviewer: ```suggestion
which type would you choose for superblock entries, and which type for block entries and why?
```

\endassignment

\solution

If a superblock spans 1600 bits, then the last block value within a superblock can at most count 1599 `1s`.
Thus, a `uint16_t` is the smallest type that is still able represent this number

> and should be preferred to larger types, to reduce the space of the block data structure.

Reviewer: ```suggestion
and should be preferred to larger types to reduce the space of the block data structure.
```

2. Add a member function `void construct(size_t const block_size = 64u, size_t const superblock_size = 512u)`
that fills the member variables `blocks` and `superblocks`.

With the following main function:

\snippet test/snippet/nanocourse/solution4.cpp main

your program should print the following:

```
Superblocks:
0 25 50 75

Blocks:
0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24
0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24
0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24
0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24
```

If you aren't experienced with C++, you can use the following backbone:

\hint

```cpp
    // The member function takes two parameters:
    // `block_size` which is defaulted to be 64 and `superblock_size` which is defaulted to 512.
    // Defaulted means that you can also call the function without parameters `construct()`.
    void construct(size_t const block_size = 64, size_t const superblock_size = 512)
    {
        size_t number_of_bits = data.size() * 64;

        // resize the two arrays of blocks and superblocks

>         blocks = std::vector<uint16_t>(/*TODO: how many blocks do I need?*/, 0);

Reviewer: Actually we could just do `blocks.resize(/*TODO: how many blocks do I need?*/)`? 
Same in the solutions.

Author: :woman_facepalming:  yes

Author: I will also do this in the solution 1 constructor instead of the direct data member initialisation to keep it simple and consistent.

\include test/snippet/nanocourse/solution4_intrinsic.cpp

If you have some time to spare, increase the size of B, and do some runtime tests for the construction.
The construction using popcount should be considerably faster.

\endsolution

### Rank queries

Now that we have the helper data structures of block and superblocks,
we will implement the actual support for rank queries.

\assignment{Rank queries}

Implement a member function

```cpp
    uint64_t rank(size_t const i) const
    {

>         // you code here

Reviewer: ```suggestion
        // your code here
```
